crates/swc_ecma_transforms_module/src/system_js.rs

```diff
@@ -0,0 +1,1224 @@
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use swc_atoms::{js_word, JsWord};
+use swc_common::{BytePos, FileName, Mark, Span, SyntaxContext, DUMMY_SP};
+use swc_ecma_ast::*;
+use swc_ecma_utils::{quote_ident, quote_str, var::VarCollector, ExprFactory};
+use swc_ecma_visit::{noop_fold_type, Fold, FoldWith, VisitWith};
+
+use crate::{
+    path::{ImportResolver, NoopImportResolver},
+    util::{local_name_for_src, use_strict},
+};
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct Config {
+    #[serde(default)]
+    pub allow_top_level_this: bool,
+}
+
+struct SystemJs<R>
+where
+    R: ImportResolver,
+{
+    root_mark: Mark,
+    resolver: Option<(R, FileName)>,
+    config: Config,
+
+    declare_var_idents: Vec<Ident>,
+    export_map: HashMap<(JsWord, Option<SyntaxContext>), Vec<JsWord>>,
+    export_names: Vec<JsWord>,
+    export_values: Vec<Expr>,
+    export_identifier: Ident,
+    context_identifier: Ident,
+    tla: bool,
+    enter_async_fn: bool,
+    enter_fn: bool,
+    root_fn_decl_idents: Vec<Ident>,
+    module_item_meta_list: Vec<ModuleItemMeta>,
+    import_ident_names: Vec<JsWord>,
+}
+
+pub fn system_js(root_mark: Mark, config: Config) -> impl Fold {
+    SystemJs {
+        root_mark,
+        resolver: None::<(NoopImportResolver, _)>,
+        config,
+
+        declare_var_idents: vec![],
+        export_map: HashMap::new(),
+        export_names: vec![],
+        export_values: vec![],
+        export_identifier: quote_ident!("_export"),
+        context_identifier: quote_ident!("_context"),
+        tla: false,
+        enter_async_fn: false,
+        enter_fn: false,
+        root_fn_decl_idents: vec![],
+        module_item_meta_list: vec![],
+        import_ident_names: vec![],
+    }
+}
+
+pub fn system_js_with_resolver<R>(
+    resolver: R,
+    base: FileName,
+    root_mark: Mark,
+    config: Config,
+) -> impl Fold
+where
+    R: ImportResolver,
+{
+    SystemJs {
+        root_mark,
+        resolver: Some((resolver, base)),
+        config,
+
+        declare_var_idents: vec![],
+        export_map: HashMap::new(),
+        export_names: vec![],
+        export_values: vec![],
+        export_identifier: quote_ident!("_export"),
+        context_identifier: quote_ident!("_context"),
+        tla: false,
+        enter_async_fn: false,
+        enter_fn: false,
+        root_fn_decl_idents: vec![],
+        module_item_meta_list: vec![],
+        import_ident_names: vec![],
+    }
+}
+
+struct ModuleItemMeta {
+    export_names: Vec<JsWord>,
+    export_values: Vec<Expr>,
+    has_export_all: bool,
+    src: JsWord,
+    stmts: Vec<Stmt>,
```

Comment:
Can you rename this field? `stmts` is too general and I'm not sure which values it store.